README.md
# HttpStreamingClient [![Coverage Status](https://coveralls.io/repos/adobe-research/http_streaming_client/badge.png)](https://coveralls.io/r/adobe-research/http_streaming_client) [![Gem Version](https://badge.fury.io/rb/http_streaming_client.png)](http://badge.fury.io/rb/http_streaming_client)

Ruby HTTP client with support for HTTP 1.1 streaming, GZIP and zlib compressed streams, and chunked transfer encoding. Includes extensible OAuth support for the following:

* Adobe Analytics Firehose
* Twitter Streaming APIs

## Ruby Version

MRI 2.1.1p76 and JRuby jruby-1.7.12. Install via rvm: https://rvm.io/

## Installation

Add this line to your application's Gemfile:

    gem 'http_streaming_client'

And then execute:

    $ bundle install

Or install it directly via gem with:

    $ gem install http_streaming_client

## Simple Example

Twitter Firehose Sample Stream:

```ruby
require 'http_streaming_client'

twitter_stream_url = "https://stream.twitter.com/1.1/statuses/sample.json"

# Generate the HMAC-SHA1 Twitter OAuth authorization header
authorization = HttpStreamingClient::Oauth::Twitter.generate_authorization(twitter_stream_url, "get", {}, OAUTH_CONSUMER_KEY, OAUTH_CONSUMER_SECRET, OAUTH_TOKEN, OAUTH_TOKEN_SECRET)

# Configure the client
client = HttpStreamingClient::Client.new(compression: true, reconnect: true, reconnect_interval: 10, reconnect_attempts: 60)

# Open the connection and start processing messages
response = client.get(twitter_stream_url, {:headers => {'Authorization' => "#{authorization}" }}) { |line|
  logger.info "Received a line that we could parse into JSON if we want: #{line}"
  client.interrupt if we_want_to_stop_receiving_messages?
}
```

For more examples, take a look at

* spec/client_spec.rb
* tools/adobe_firehose.rb
* tools/adobe_firehose_performance_test.rb
* tools/twitter_firehose.rb
* tools/twitter_firehose_performance_test.rb

## Client Configuration Options

The following options are supported as hash option parameters for HttpStreamingClient::Client.new:

GZip Compression

    compression: true/false (default: true)

Automatic Socket Reconnect Functions

    reconnect: true/false (default: false)

Reconnect Interval

    reconnect_interval: interval_seconds (default: 1 second)

Maximum Reconnect Attempts

    reconnect_attempts: num_attempts (default: 10)

## Logging

HTTP protocol trace logging is available as :debug level logging. The gem supports configurable logging to both STDOUT and a log file, and includes a Railtie to use Rails.logger when the gem is included in a Rails application.

To configure gem logging to STDOUT, specify the following in your code:

    HttpStreamingClient.logger.console = true

To configure gem logging to a log file named "test.log", specify the following in your code:

    HttpStreamingClient.logger.logfile = true

And to set the log level, specify the following in your code (e.g. to set the log level to :debug):

    HttpStreamingClient.logger.level = Logger::DEBUG

## Streaming Service Credentials

The command line tools for the Adobe Analytics Firehose and Twitter's Streaming APIs require valid Adobe and Twitter credentials. Unit tests execute against the Twitter sample firehose and the Adobe Analytics Firehose and also require valid service credentials.

To configure Adobe Analytics Firehose credentials, copy lib/http_streaming_client/credentials/adobe.rb.sample to lib/http_streaming_client/credentials/adobe.rb, and edit the file to include valid Adobe Analytics Firehose API credentials.

To configure Twitter credentials, copy lib/http_streaming_client/credentials/twitter.rb.sample to lib/http_streaming_client/credentials/twitter.rb, and edit the file to include valid Twitter API credentials.

## Command Line Tools

To run the sample Adobe Analytics Firehose client, execute the following after configuring valid service credentials:

    $ ruby tools/adobe_firehose.rb

To run the Adobe Analytics Firehose performance test client, execute the following after configuring valid service credentials:

    $ ruby tools/adobe_firehose_performance_test.rb

The performance test client will emit performance test metrics to stdout as JSON objects with the following format:

    {"total_records_received":"22000","total_elapsed_time_sec":"118.98","total_records_per_sec":"184.9","total_kbytes_per_sec":"307.5","interval_records_received":1000,"interval_elapsed_time_sec":"5.0","interval_records_per_sec":"200.15","interval_kbytes_per_sec":"331.89"}

The batch size per metric output and the total number of records in the performance test run can be configured within the script.

To run the sample Twitter Firehose client, execute the following after configuring valid service credentials:

    $ ruby tools/twitter_firehose.rb

To run the Twitter Firehose performance test client, execute the following after configuring valid service credentials:

    $ ruby tools/twitter_firehose_performance_test.rb

The performance metrics output and tunable settings in the script are identical to those for the Adobe Analytics Firehose performance test client.

All tools emit JSON object streams to stdout.

## Unit Test Coverage

Unit test suite implemented with rspec and simplecov. Run via:

    $ rake
or:

    $ rspec

Individual test suites in the spec directory can be run via:

    $ rspec spec/<spec filename>.spec

An HTML coverage report is generated at the end of a full test run in the coverage directory.

## Fixed Issues

* See [![CHANGELOG](CHANGELOG)](CHANGELOG)

## License

Licensed under the Apache Software License 2.0. See [![LICENSE](LICENSE)](LICENSE) file.


CHANGELOG
http_streaming_client 0.9.6 (07.21.2014)
========================================
* Fixed bug with connection close and uncompressed chunked transfer encoding

http_streaming_client 0.9.5 (05.23.2014)
========================================
* Added logging tag functions to add tags to log messages for multi-threaded environments

http_streaming_client 0.9.4 (05.22.2014)
========================================
* Fixed bug with reconnect management and lazy authorization header generation
* Tested and updated for MRI ruby-2.1.1p76 and JRuby jruby-1.7.12

http_streaming_client 0.9.3 (05.09.2014)
========================================
* Critical bug fix for GZip chunked deceompression handling

http_streaming_client 0.9.1 (04.30.2014)
========================================
* added support for dynamic HTTP header and client option generation via a duck-typed factory parameter

http_streaming_client 0.9.0 (04.28.2014)
========================================
* Tested and fixed for MRI ruby-2.0.0-p451 and JRuby jruby-1.7.12
* Fixed chunked encoding support handler, support for gzip'd blocks spanning chunks

http_streaming_client 0.8.11 ()
========================================
* Fixed warn logging for non-200 HTTP response codes

http_streaming_client 0.8.10 (1.27.2014)
========================================
* bug fix for chunked transfer encoding response with no compression and no block handler

http_streaming_client 0.8.9 (1.26.2014)
========================================
* A few minor logging warn/debug changes

http_streaming_client 0.8.8 (1.7.2014)
========================================
* Fixed Issue #1, memory leak due to accumulated messages within the buffer pipeline
* Added automatic reconnection support and configuration options
* Added automatic HTTP 301 redirect support
* Added new unit tests for reconnect and for memory leak detection
* Updated performance tests

http_streaming_client 0.8.5 (12.10.2013)
========================================
* Updates to support MRI ruby 2.0.0-p353

http_streaming_client 0.8.0 (11.09.2013)
========================================
* added adobe firehose performance test client
* improved error handling for adobe firehose client
* added Railtie for Rails logger integration

http_streaming_client 0.5.0 (10.25.2013)
========================================
* Added support for the Adobe Analytics Firehose
* Refactored OAuth support to be extensible and support Adobe and Twitter mechanisms
* Refactored streaming support to support both chunked transfer encoding and line-oriented HTTP 1.1 streaming
* Refactored decompression support to support both compressed chunked transfer encoding and compressed line-oriented HTTP 1.1 streaming
* Added Adobe Firehose and Twitter Firehose command line scripts
* Added unit test coverage reports with simplecov

http_streaming_client 0.1.0 (10.23.2013)
========================================
* initial commit
* HTTP client with streaming support for GZIP compressed streams and chunked transfer encoding
* unit test suite


lib/http_streaming_client/client.rb
###########################################################################
##
## http_streaming_client
##
## Ruby HTTP client with support for HTTP 1.1 streaming, GZIP compressed
## streams, and chunked transfer encoding. Includes extensible OAuth
## support for the Adobe Analytics Firehose and Twitter Streaming APIs.
##
## David Tompkins -- 11/8/2013
## tompkins@adobe_dot_com
##
###########################################################################
##
## Copyright (c) 2013 Adobe Systems Incorporated. All rights reserved.
##
## Licensed under the Apache License, Version 2.0 (the "License");
## you may not use this file except in compliance with the License.
## You may obtain a copy of the License at
##
## http://www.apache.org/licenses/LICENSE-2.0
##
## Unless required by applicable law or agreed to in writing, software
## distributed under the License is distributed on an "AS IS" BASIS,
## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
## See the License for the specific language governing permissions and
## limitations under the License.
##
###########################################################################

require 'socket'
require 'uri'
require 'zlib'

require "http_streaming_client/version"
require "http_streaming_client/custom_logger"
require "http_streaming_client/errors"
require "http_streaming_client/decoders/gzip"
require "http_streaming_client/decoders/chunked"

module HttpStreamingClient

  class Client

    attr_accessor :socket, :interrupted, :compression_requested, :reconnect_requested, :reconnect_attempts, :reconnect_interval

    ALLOWED_MIME_TYPES = ["application/json", "text/plain", "text/html"]

    MAX_RECONNECT_ATTEMPTS = 10
    RECONNECT_INTERVAL_SEC = 1

    def self.logger
      HttpStreamingClient.logger
    end

    def logger
      HttpStreamingClient.logger
    end

    def initialize(opts = {})
      logger.debug("Client.new: #{opts}")
      @socket = nil
      @interrupted = false

      @compression_requested = opts[:compression].nil? ? true : opts[:compression]
      logger.debug("compression is #{@compression_requested}")

      @reconnect_requested = opts[:reconnect].nil? ? false : opts[:reconnect]
      logger.debug("reconnect is #{@reconnect_requested}")

      if @reconnect_requested then
	@reconnect_count = 0

	@reconnect_attempts = opts[:reconnect_attempts].nil? ? MAX_RECONNECT_ATTEMPTS : opts[:reconnect_attempts]
	logger.debug("reconnect_attempts is #{@reconnect_attempts}")

	@reconnect_interval = opts[:reconnect_interval].nil? ? RECONNECT_INTERVAL_SEC : opts[:reconnect_interval]
	logger.debug("reconnect_interval is #{@reconnect_interval}")

      end

    end

    def self.get(uri, opts = {}, &block)
      logger.debug("get:#{uri}")
      self.new.request("GET", uri, opts, &block)
    end

    def get(uri, opts = {}, &block)
      logger.debug("get(interrupt):#{uri}")
      @interrupted = false
      begin
	request("GET", uri, opts, &block)
      rescue IOError => e
	raise e unless @interrupted
      end
    end

    def self.post(uri, body, opts = {}, &block)
      logger.debug("post:#{uri}")
      self.new.request("POST", uri, opts.merge({:body => body}), &block)
    end

    def post(uri, body, opts = {}, &block)
      logger.debug("post(interrupt):#{uri}")
      @interrupted = false
      begin
	request("POST", uri, opts.merge({:body => body}), &block)
      rescue IOError => e
	raise e unless @interrupted
      end
    end

    def interrupt
      logger.debug("interrupt")
      @interrupted = true
      @socket.close unless @socket.nil?
    end

    def request(method, uri, opts = {}, &block)

      logger.debug("Client::request:#{method}:#{uri}:#{opts}")

      if uri.is_a?(String)
	uri = URI.parse(uri)
      end

      @reconnect_count = 0 if @reconnect_requested
      @options_factory = opts.delete(:options_factory)
      @static_body_option = opts[:body]

      begin

	opts[:body] = @static_body_option unless @static_body_option.nil?

	if !@options_factory.nil? then
	  if @options_factory.respond_to? "get_options" then
	    logger.debug("Client::request:options_factory detected")
	    generated_options = @options_factory.get_options
	    logger.debug("Client::request:options_factory:#{generated_options}")
	    opts.merge!(generated_options || {})
	  else
	    logger.warn("Client::request:options_factory detected, but does not respond to get_options(). Ignoring.")
	  end
	end

	default_headers = {
	  "User-Agent" => opts["User-Agent"] || "HttpStreamingClient #{HttpStreamingClient::VERSION}",
	  "Accept" => "*/*",
	  "Accept-Charset" => "utf-8"
	}

	if method == "POST" || method == "PUT"
	  default_headers["Content-Type"] = opts["Content-Type"] || "application/x-www-form-urlencoded;charset=UTF-8"
	  body = opts.delete(:body)
	  if body.is_a?(Hash)
	    body = body.keys.collect {|param| "#{URI.escape(param.to_s)}=#{URI.escape(body[param].to_s)}"}.join('&')
	  end
	  default_headers["Content-Length"] = body.length
	end

	unless uri.userinfo.nil?
	  default_headers["Authorization"] = "Basic #{[uri.userinfo].pack('m').strip!}\r\n"
	end

	encodings = []
	encodings << "gzip" if (@compression_requested and opts[:compression].nil?) or opts[:compression]
	if encodings.any?
	  default_headers["Accept-Encoding"] = "#{encodings.join(',')}"
	end

	headers = default_headers.merge(opts[:headers] || {})
	logger.debug "request headers: #{headers}"

	socket = initialize_socket(uri, opts)
	logger.info("Connected to #{uri}")

	request = "#{method} #{uri.path}#{uri.query ? "?"+uri.query : nil} HTTP/1.1\r\n"
	request << "Host: #{uri.host}\r\n"
	headers.each do |k, v|
	  request << "#{k}: #{v}\r\n"
	end
	request << "\r\n"
	if method == "POST"
	  request << body
	end

	socket.write(request)

	response_head = {}
	response_head[:headers] = {}

	socket.each_line do |line|
	  if line == "\r\n" then
	    break
	  else
	    header = line.split(": ")
	    if header.size == 1
	      header = header[0].split(" ")
	      response_head[:version] = header[0]
	      response_head[:code] = header[1].to_i
	      response_head[:msg] = header[2]
	      logger.debug "HTTP response code is #{response_head[:code]}"
	    else
	      response_head[:headers][camelize_header_name(header[0])] = header[1].strip
	    end
	  end
	end

	logger.debug "response headers:#{response_head[:headers]}"

	if response_head[:code] == 301 then
	  location = response_head[:headers]["Location"]
	  raise InvalidRedirect, "Unable to find Location header for HTTP 301 response" if location.nil?
	  logger.warn "Received HTTP 301 redirect to #{location}, following..."
	  socket.close if !socket.nil? and !socket.closed?
	  opts.delete(:socket)
	  return request(method, location, opts, &block)
	end

	content_length = response_head[:headers]["Content-Length"].to_i
	logger.debug "content-length: #{content_length}"

	content_type = response_head[:headers]["Content-Type"].split(';').first
	logger.debug "content-type: #{content_type}"

	response_compression = false

	if ALLOWED_MIME_TYPES.include?(content_type)
	  case response_head[:headers]["Content-Encoding"]
	  when "gzip"
	    response_compression = true
	  end
	else
	  raise InvalidContentType, "invalid response MIME type: #{content_type}"
	end

	if (response_head[:code] != 200)
	  s = "Received HTTP #{response_head[:code]} response"
	  logger.debug "request: #{request}"
	  response = socket.read(content_length)
	  logger.debug "response: #{response}"
	  raise HttpError.new(response_head[:code], "Received HTTP #{response_head[:code]} response", response_head[:headers], response)
	end

	if response_head[:headers]["Transfer-Encoding"] == 'chunked'
	  partial = nil
	  decoder = nil
	  response = ""

	  if response_compression then
	    logger.debug "chunked transfer encoding with compression detected"
	    if block_given? then
	      decoder = HttpStreamingClient::Decoders::GZip.new { |line|
		logger.debug "read #{line.size} uncompressed bytes, decoder queue bytes:#{decoder.size}"
		block.call(line) unless @interrupted }
	    else
	      decoder = HttpStreamingClient::Decoders::GZip.new { |line|
		logger.debug "read #{line.size} uncompressed bytes, #{response.size} bytes total, decoder queue bytes:#{decoder.size}"
		response << line unless @interrupted }
	    end
	  else
	    logger.debug "chunked transfer encoding with no compression detected"
	    if block_given? then
	      decoder = HttpStreamingClient::Decoders::Chunked.new { |line|
		logger.debug "read #{line.size} uncompressed bytes, decoder queue bytes:#{decoder.size}"
		block.call(line) unless @interrupted }
	    else
	      decoder = HttpStreamingClient::Decoders::Chunked.new { |line|
		logger.debug "read #{line.size} uncompressed bytes, #{response.size} bytes total, decoder queue bytes:#{decoder.size}"
		response << line unless @interrupted }
	    end
	  end

	  while !socket.eof? && (line = socket.gets)
	    chunkLeft = 0

	    if line.match /^0\r\n/ then
	      logger.debug "received zero length chunk, chunked encoding EOF"
	      logger.debug "EOF line: #{line}"
	      break
	    end

	    next if line == "\r\n"

	    size = line.hex
	    logger.debug "chunk size:#{size}"

	    partial = socket.read(size)
	    next if partial.nil?

	    remaining = size-partial.size
	    logger.debug "read #{partial.size} bytes, #{remaining} bytes remaining"
	    until remaining == 0
	      partial << socket.read(remaining)
	      remaining = size-partial.size
	      logger.debug "read #{partial.size} bytes, #{remaining} bytes remaining"
	    end

	    decoder << partial

	    if !block_given? then
	      logger.debug "no block specified, returning chunk results and halting streaming response"
	      # flush any remaining uncompressed chunked decoder bytes into response by pushing a newline
	      decoder << "\n" if !response_compression and decoder.size > 0
	      return response
	    else
	      if @interrupted then
		return if response_compression
	        # flush any remaining uncompressed chunked decoder bytes into response by pushing a newline
	        decoder << "\n" if !response_compression and decoder.size > 0
		return response
	      end
	    end

	  end

	  logger.debug "socket EOF detected" if socket.eof?
	  raise ReconnectRequest if @reconnect_requested
          # flush any remaining uncompressed chunked decoder bytes into response by pushing a newline
          decoder << "\n" if !response_compression and decoder.size > 0
	  return response

	else
	  # Not chunked transfer encoding, but potentially gzip'd, and potentially streaming with content-length = 0

	  if content_length > 0 then
	    bits = socket.read(content_length)
	    logger.debug "read #{content_length} bytes"
	    return bits if !response_compression
	    logger.debug "response compression detected"
	    begin
	      decoder = Zlib::GzipReader.new(StringIO.new(bits))
	      return decoder.read
	    rescue Zlib::Error
	      raise DecoderError
	    end
	  end

	  if response_compression then

	    logger.debug "response compression detected"
	    decoder = nil
	    response = ""

	    if block_given? then
	      decoder = HttpStreamingClient::Decoders::GZip.new { |line|
		logger.debug "read #{line.size} uncompressed bytes, decoder queue bytes:#{decoder.size}"
		block.call(line) unless @interrupted }
	    else
	      decoder = HttpStreamingClient::Decoders::GZip.new { |line|
		logger.debug "read #{line.size} uncompressed bytes, #{response.size} bytes total, decoder queue bytes:#{decoder.size}"
		response << line unless @interrupted }
	    end

	    while (!socket.eof? and !(line = socket.read_nonblock(2048)).nil?)
	      logger.debug "read compressed line, #{line.size} bytes"
	      decoder << line
	      break response if @interrupted
	    end

	    logger.debug "EOF detected"
	    raise ReconnectRequest if @reconnect_requested
	    return response

	  else

	    response = ""

	    while (!socket.eof? and !(line = socket.readline).nil?)
	      if block_given? then
		yield line
		logger.debug "read #{line.size} bytes"
	      else
		logger.debug "read #{line.size} bytes, #{response.size} bytes total"
		response << line
	      end
	      break if @interrupted
	    end

	    raise ReconnectRequest if @reconnect_requested
	    return response

	  end
	end
      rescue => e
	return if @interrupted
	logger.error "Error Detected: #{e}" unless e.instance_of? ReconnectRequest
	decoder.close if !decoder.nil?
	socket.close if !socket.nil? and !socket.closed?
	opts.delete(:socket)

	if @reconnect_requested then
	  logger.info "Connection closed. Reconnect requested. Trying..."
	  @reconnect_count = @reconnect_count + 1
	  logger.info "Reconnect attempt #{@reconnect_count} of #{@reconnect_attempts}, sleeping for #{@reconnect_interval} seconds..."
	  sleep @reconnect_interval
	  retry if @reconnect_count < @reconnect_attempts
	  logger.info "Maximum number of failed reconnect attempts reached (#{@reconnect_attempts}). Exiting."
	end

	raise e unless e.instance_of? ReconnectRequest
      end
    ensure
      logger.debug "ensure socket closed"
      decoder.close if !decoder.nil?
      socket.close if !socket.nil? and !socket.closed?
      opts.delete(:socket)
    end

    private

    def camelize_header_name(header_name)
      (header_name.split('-').map {|s| s.capitalize}).join('-')
    end

    def initialize_socket(uri, opts = {})
      return opts[:socket] if opts[:socket]

      if uri.is_a?(String)
	uri = URI.parse(uri)
      end

      logger.debug "Connecting to #{uri.host}:#{uri.port}..."

      @socket = TCPSocket.new(uri.host, uri.port)

      if uri.scheme.eql? "https"
	ctx = OpenSSL::SSL::SSLContext.new
	ctx.set_params(verify_mode: OpenSSL::SSL::VERIFY_PEER)
	@socket = OpenSSL::SSL::SSLSocket.new(@socket, ctx).tap do |socket|
	  socket.sync_close = true
	  socket.connect
	end
      end

      logger.debug "Connected to #{uri.host}:#{uri.port}"
      opts.merge!({:socket => @socket})
      @interrupted = false
      return opts[:socket]
    end
  end

end
